connectors/dynamodb/stream/helper.go

```diff
@@ -0,0 +1,350 @@
+				select {
+				case <-ctx.Done():
+					close(requestCh)
+					<-done
+					return ctx.Err()
+				case requester := <-s.shardRequestCh:
+					currentRequests = append(currentRequests, requester)
+				}
+			}
+			requestCh <- currentRequests
+			currentRequests = nil
+		case requester := <-s.shardRequestCh:
+			currentRequests = append(currentRequests, requester)
+		case <-ctx.Done():
+			close(requestCh)
+			<-done
+			return ctx.Err()
+		}
+	}
+}
+
+func (s *dynamodbStreamHelper) shardIteratorInputForEarliest(ctx context.Context, arn string) ([]*dynamodbstreams.GetShardIteratorInput, error) {
+	var res []*dynamodbstreams.GetShardIteratorInput
+	shards, err := s.GetShards(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	shardSet := map[string]struct{}{}
+	var candidates []types.Shard
+
+	for _, shard := range shards {
+		shardSet[*shard.ShardId] = struct{}{}
+	}
+	for _, shard := range shards {
+		if shard.ParentShardId == nil {
+			candidates = append(candidates, shard)
+		} else if _, ok := shardSet[*shard.ParentShardId]; !ok {
+			candidates = append(candidates, shard)
+		}
+	}
+	for _, shard := range candidates {
+		res = append(res, &dynamodbstreams.GetShardIteratorInput{
+			ShardId:           shard.ShardId,
+			ShardIteratorType: types.ShardIteratorTypeTrimHorizon,
+			StreamArn:         aws.String(arn),
+		})
+	}
+	return res, nil
+}
+
+func (s *dynamodbStreamHelper) shardIteratorInputForLatest(ctx context.Context, arn string) ([]*dynamodbstreams.GetShardIteratorInput, error) {
+	var res []*dynamodbstreams.GetShardIteratorInput
+	shards, err := s.GetShards(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var openShards []types.Shard
+	for _, shard := range shards {
+		if shard.SequenceNumberRange.EndingSequenceNumber != nil {
+			continue
+		}
+		openShards = append(openShards, shard)
+	}
+	for _, shard := range openShards {
+		res = append(res, &dynamodbstreams.GetShardIteratorInput{
+			ShardId:           shard.ShardId,
+			ShardIteratorType: types.ShardIteratorTypeLatest,
+			StreamArn:         aws.String(arn),
+		})
+	}
+	return res, nil
+}
+
+func (s *dynamodbStreamHelper) approximateLatestState(ctx context.Context, approximateNow time.Time) (StreamState, error) {
+	// Strategy here is to read only from the open shards, and find the sequence number just before the approximate time.
+	// We cannot use the LATEST shard iterator since that relies on new entries
+	var openShards []types.Shard
+	shards, err := s.GetShards(ctx)
+	if err != nil {
+		return StreamState{}, err
+	}
+	for _, shard := range shards {
+		if shard.SequenceNumberRange.EndingSequenceNumber != nil {
+			continue
+		}
+		openShards = append(openShards, shard)
+	}
+
+	streamARN := aws.String(s.streamARN)
+	state := NewStreamState(s.streamARN)
+
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	for _, shard := range openShards {
+		eg.Go(func() error {
+			shardIteratorRes, err := s.streamClient.GetShardIterator(egCtx, &dynamodbstreams.GetShardIteratorInput{
```

Comment:
when does this start reading? At the beginning of time?

Reply:
it starts reading from the earliest entries in the shards that are still open (no ending sequence number)